### PR TITLE
Fixed race condition, updated default masquerades

### PR DESCRIPTION
**Pull request by oxtoacart**
For getlantern/lantern-internal#995

### CODE REVIEW

File: masquerade.go [hunk 1]
```diff
@@ -24,11 +23,19 @@ type Masquerade struct {
 	// IpAddress: pre-resolved ip address to use instead of Domain (if
 	// available)
 	IpAddress string
+}
+
+type masquerade struct {
```
Comment (oxtoacart) [inline masquerade.go]: The crux of the fix is that we copy the provided Masquerade into an internal data structure, so whoever supplied us masquerades can do what they want with them without having to worry about synchronizing.

File: cache.go [hunk 2]
```diff
@@ -70,11 +70,13 @@ func (d *direct) fillCache(cache []*Masquerade, cacheFile string) {
 			log.Debug("Saving updated masquerade cache")
 			// Truncate cache to max length if necessary
 			if len(cache) > d.maxCacheSize {
-				truncated := make([]*Masquerade, d.maxCacheSize)
+				truncated := make([]*masquerade, d.maxCacheSize)
 				copy(truncated, cache[len(cache)-d.maxCacheSize:])
 				cache = truncated
 			}
+			d.mx.RLock()
```
Comment (oxtoacart) [inline cache.go]: We now just use a single mutex for the whole direct struct, which we lock when marshaling the masquerades into the cache.

File: masquerade.go (hunk 1 above)
Comment (joesis) [reply masquerade.go]: And simply store and pass the object by value to reduce race condition

File: cache.go (hunk 2 above)
Comment (joesis) [reply cache.go]: Actually, always passing masquerade by value in the channels would eliminate locking at all.

Comment (oxtoacart) [reply cache.go]: Yeah, but it adds extra copy steps.

**Comment by oxtoacart** [comment]
@joesis Thanks for the review! Unless you feel strongly about eliminating locking, I'd say we leave as-is since it works.

**Comment by joesis** [comment]
@oxtoacart I just felt that it's a bit counter-intuitive why modifying one masquerade would affect the cache, and adding a lock to two logically irrelevant parts may be hard to understand in the future. I added benchmarks in issue995-joesis branch, and yes, copying value is 10+ times slower than copying pointer, but when it comes to channel operation, the performance is comparable.

```
BenchmarkCopyPointer-4      	 5000000	       249 ns/op
BenchmarkCopyValue-4        	  500000	      2438 ns/op
BenchmarkAssignPointer-4    	 3000000	       439 ns/op
BenchmarkAssignValue-4      	  200000	      6785 ns/op
BenchmarkChannelPointer-4   	   10000	    147437 ns/op
BenchmarkChannelValue-4     	   10000	    163833 ns/op
```

**Comment by oxtoacart** [comment]
@joesis If you have a working version that uses a channels of values I'm happy to use that instead. Performance wasn't really my concern so much as just not spending more time on this since we already have something that works :)

**Comment by oxtoacart** [comment]
BTW @joesis, I do think that your value-based channels approach is preferable from a design perspective.

**Pull request merged**